File: 75-SortColors.java
//note that this solution is essentially the partition() method in quicksort, applied on both left and right side at the same time
public class Solution {
    public void sortColors(int[] nums) {
        int red=0;
        int blue=nums.length-1;
        for(int i=0; i<=blue && i < nums.length;) {
            if (nums[i] == 0 && i > red ) {
                swap(nums,i,red);
                red++;
            } else if (nums[i] == 2 && i < blue){
                swap(nums,i,blue);
                blue--;
            }
            else
                i++;
        }
    }
    
    private void swap(int nums[], int a, int b){
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}

File: cpp/75-SortColors.cpp
class Solution {
public:
    void swap(vector<int>& nums, int p1, int p2){
        int tmp = nums[p1];
        nums[p1] = nums[p2];
        nums[p2] = tmp;
    }

    void sortColors(vector<int>& nums) {
        if (nums.size() == 0){
            return;
        }
        int p1 = 0, p2 = nums.size() - 1, curr = 0;
        while (p1 <= p2 && curr <= p2){
            if (nums[curr] == 0){
                swap(nums, curr, p1);
                p1++;
                curr++;
            }else if (nums[curr] == 1){
                curr++;
            }else{
                swap(nums, curr, p2);
                //attention here!
                //no curr++;
                p2--;
            }
        }
    }
};
